keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1363 @@
+
+```cel
+Object{
+  metadata: Object.metadata{
+    annotations:
+      ?"some-annotation": optional.none()
+  }
+  spec: Object.spec{
+    someField: object.annotations["some-annotation"]
+  }
+}
+```
+
+#### Use Case - mutation VS controller fight
+https://github.com/open-policy-agent/gatekeeper/issues/2963#issuecomment-1683971371
+Out of scope. The proposed feature is going to be added as an admission plugin which has no control over other controllers potentially being added.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Risk: Enabling CEL object instantiation enables users to allocate memory in
+a more directly way than previously available.
+
+- Mitigation/Justification: List and map literals were already possible, so this
+  doesn't fundamentally change the memory allocation situation. We could further
+  mitigate by statically estimating the "memory cost" of CEL expressions that
+  include any form of data literal (list, map, object or scalar).
+
+Risk: Final state of the object might not match the output of mutation policies.
+There might be multiple mutating admission policies, mutating webhooks, other controllers trying to mutate the incoming request and each happens separately, 
+and they might mutate the same part of the object. It might be hard to ensure that the final state matches expectations. 
+
+- Mitigation/Justification: For best practice, the validation process is highly recommended whenever there is a mutation process set up. 
+The validating admission policy is recommended to be set up whenever a mutation admission policy is set to verify the final state of the data matches the expectation. Also refer to the Safety section for further details.
+
+
+
+## Design Details
+
+### Object type names
+
+As part of this enhancement, we are enabling CEL object instantiation, which
+we have left disabled in previous CEL features.
+
+When enabling CEL object instantiation we need to decide:
+
+- How object type names will represented in CEL. This KEP shows a
+  "Object.spec.container" naming system. Is this what we will use? Or will be
+  use actual schema type names, e.g. "v1.Pod.spec.container"?
+- Will validating admission features also gain the ability to instantiate CEL
+  types? This has memory consumption implications.
+
+### SSA Merge algorithm reuse
+
+Reusing Server Side Apply merge algorithms is complicated by presence of
+numerous different representations of schema types in Kubernetes (Structural
+schemas, multiple OpenAPI schema representations, SMD schemas, ...). For an
+initial alpha, we may simply perform the needed conversions, but longer term
+memory consumption and runtime performance may demand that we minimize the
+conversions needed.
```

Comment:
I think it should definitely be possible. As long as we have a straightforward way of creating patches out of the proposed syntax. For some of the field removals, there may need to be two patches done (one to take ownership of the field, and another to clear it). 

SSA merge requires a TypeResolver. It is easy to create one from a map of kube-openapi definitions so for alpha that definitely can be used. I anticipate being able to share the TypeResolver between the admission plugin and rest of the apiserver would be hard yet possible; but not sure if we necessarily want that (do any of our other admission plugins share resources with the apiserver that way? Just thinking out loud)

One of the big nagging questions I think left out here for me is what field manager is going to be used and how will we address conflicts? Will all field managers force conflicts? Will all mutating admission policies use the same field manager? 

I think the design of SSA forces you to use a different field manager per policy-binding-param. Since if they shared a field manager then they would keep reverting each other's changes.

Reply:
Thanks for the comment! For the MAP implementation we properly would go with type converter as needed. I agree that sharing TypeResolver between admission and apiserver is tricky and it is not what proposed here. Here is notes on future possible simplification including having as less representation of schema as possible(or try to have a general representation of schema used in the future) :)